cmd/geth/dbcmd.go

```diff
@@ -507,3 +544,117 @@ func freezerInspect(ctx *cli.Context) error {
 	}
 	return nil
 }
+
+// importFuncs defines all supported chain data import functions.
+var importFuncs = map[string]func(db ethdb.Database, fn string, interrupt chan struct{}) error{
+	"preimage": utils.ImportPreimages,
+	"snapshot": utils.ImportSnapshot,
+}
+
+func importChaindata(ctx *cli.Context) error {
+	if ctx.NArg() < 2 {
+		return fmt.Errorf("required arguments: %v", ctx.Command.ArgsUsage)
+	}
+	// Parse the required chain data type, make sure it's supported.
+	kind := ctx.Args().Get(0)
+	kind = strings.ToLower(strings.Trim(kind, " "))
+	fn, ok := importFuncs[kind]
+	if !ok {
+		var kinds []string
+		for key := range importFuncs {
+			kinds = append(kinds, key)
+		}
+		return fmt.Errorf("invalid data type %s, supported types: %s", kind, strings.Join(kinds, ", "))
+	}
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	var (
+		interrupted bool
+		interrupt   = make(chan struct{})
+	)
+	go func() {
+		sigc := make(chan os.Signal, 1)
+		signal.Notify(sigc, syscall.SIGINT, syscall.SIGTERM)
+		defer signal.Stop(sigc)
+
+		for {
+			<-sigc
+			if interrupted {
+				interrupted = true
+				close(interrupt)
+			}
+			log.Info("Got interrupt, shutting down...")
+		}
+	}()
+	db := utils.MakeChainDatabase(ctx, stack, false)
+	return fn(db, ctx.Args().Get(1), interrupt)
+}
+
+// exporter defines all the necessary information for chain data export.
+type exporter struct {
+	encoder  func(key []byte, val []byte) []byte
+	prefixes [][]byte
+}
+
+// chainExporters defines the export scheme for all exportable chain data.
+var chainExporters = map[string]*exporter{
+	"preimage": {
+		encoder: func(key []byte, val []byte) []byte {
+			return val // The key can be derived by keccak56(val).
+		},
+		prefixes: [][]byte{rawdb.PreimagePrefix},
+	},
+	"snapshot": {
+		encoder: func(key []byte, val []byte) []byte {
+			// The prefix used to identify the snapshot data type,
+			// 0: account snapshot
+			// 1: storage snapshot
```

Comment:
If we do it this way, we're limiting imports to be only the ones that are predefined. If we instead encode it as `[key, value], [key, value], [key, valiue]`, then the importer doesn't need to be datatype-aware, and we can use the same import function regardless if we're importing snapshot data or e.g. preimage data or trie data.